python/lsst/daf/butler/_query.py

```diff
@@ -0,0 +1,267 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Query",)
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, Mapping
+
+if TYPE_CHECKING:
+    from ._query_results import DataCoordinateQueryResults, DatasetQueryResults, DimensionRecordQueryResults
+    from .dimensions import DataId, DimensionElement, DimensionGroup
+    from .registry._registry import CollectionArgType
+
+
+class Query(ABC):
```

Comment:
As I mentioned at the meeting, I think we'll want to make this a concrete class with a polymorphic implementation object of some kind that's specialized on direct vs. remote butler; something like:
```
class Query:
    _relation_tree: StuffJimIsPrototypingNow  # description of the query to run, so far
    _result_rows: list[dict[ColumnTag, Any]] | None  # results from running the query, if we have them
    _backend: QueryBackend  # thing that knows how to execute _relation_tree and get back _result_rows
```
The concrete `Query` methods would then frequently return a new instance with a modified copy of `_relation_tree` and the same `_backend`, which and that wouldn't depend on whether this is a remote or direct butler.

Given that we don't actually have `StuffJimIsPrototypingNow` or a sense for what the `QueryBackend` interface would be, it might be premature to make that switch.  But I do think we'll want eventually to avoid a lot of duplication.